pkg/webhook/admission/validator.go

```diff
@@ -27,11 +28,25 @@ import (
 )
 
 // Validator defines functions for validating an operation.
+// The custom resource kind which implements this interface can validate itself.
+// To validate the custom resource with another specific struct, use CustomValidator instead.
 type Validator interface {
 	runtime.Object
-	ValidateCreate() error
-	ValidateUpdate(old runtime.Object) error
-	ValidateDelete() error
+
+	// ValidateCreate validates the object on creation.
+	// The optional warnings will be added to the response as warning messages.
+	// Return an error if the object is invalid.
+	ValidateCreate() (warnings []string, err error)
+
+	// ValidateUpdate validates the object on update. The oldObj is the object before the update.
+	// The optional warnings will be added to the response as warning messages.
+	// Return an error if the object is invalid.
+	ValidateUpdate(old runtime.Object) (warnings []string, err error)
+
+	// ValidateDelete validates the object on deletion.
+	// The optional warnings will be added to the response as warning messages.
+	// Return an error if the object is invalid.
+	ValidateDelete() (warnings []string, err error)
```

Comment:
Returning a slice and an error seems a bit error prone and not super clear, unless one goes to read the interface.

Could we move into a direction where these return a simple struct and an error instead? Something like:

```golang
type ValidationResult struct {
  Warnings []string
}
```

That way, in the future we might be able to change the signature of the struct, or add new fields without changing the interface.

Reply:
Wondering if the error should then also be part of the struct

Reply:
If `Warnings` was also based on the `error` type, couldn't it still use `field.ErrorList` making the usage of the two consistent?

Add a new helper to "k8s.io/apimachinery/pkg/api/errors": `NewWarning` and it could look like:

```
	ValidateCreate() (warnings error, err error)
```

Then the instantiation could be something like:

```
func (r *Cluster) ValidateCreate() (error, error) {
	var allWarnings field.ErrorList
	var allErrors field.ErrorList
	...
	if len(allWarnings) == 0 && len(allErrors) == 0 {
		return nil, nil
	}
	if len(allErrors == 0) {
		return apierrros.NewWarning(r.GroupVersionKind().GroupKind(), r.Name, allWarnings), nil
	}
	return apierrros.NewWarning(r.GroupVersionKind().GroupKind(), r.Name, allWarnings),
		apierrros.NewInvalid(r.GroupVersionKind().GroupKind(), r.Name, allErrors)
```

Reply:
I missed Vince's comment. Using the struct is a better idea but I still wonder if it might be good to take advantage of the pattern that's already used for building the return.

Reply:
If `apierrors.NewWarning` is a thing, we should probably use that?

Reply:
It's not, but I was suggesting it should be. :)

Reply:
one more thing, should the method return `(result ValidationResult, err error)` or `(result *ValidationResult, err error)`? 

What do you think about it ? @vincepri

Reply:
Hi @vincepri , any sugesstions? :heart:

Reply:
@STRRL I think non-pointer is the way to go. Seems to be what we do around here :) (e.g. when returning a Response in defaulter_custom.go)

I think we should make error part of the ValidationResult. Returning ValidatonResult and error separately makes it look like the error is not part of the validation result, which I think is not correct. Returning an error is basically the mainstream use case for return a validation result.

Reply:
I am not sure about making `error` as a part of `ValidationResult` is a good way. Personally, I still prefer to return them separately. 

> I think we should make error part of the ValidationResult struct though. Returning ValidatonResult and error separately makes it look like the error is not part of the validation result, which I think is not correct. Returning an error is basically the mainstream use case for returning a validation result.

IMO, "making error as a part of `ValidatinResult`" would introduce complexity into the logic, though I agree that it should be a part of the result of validating process.

In the former API, the returned `error` might represent:

- Invalid value inspected by the validation logic. Or,
- Something unexpected happens during validating process

We did not distinguish them before, and there would be a trend when developers migrate logic, they would still use `error` to represent invalid values.

Reply:
> In the former API, the returned `error` might represent:
> 
> * Invalid value inspected by the validation logic. Or,
> * Something unexpected happens during validating process
> 
> We did not distinguish them before.

To distinguish them,  maybe https://pkg.go.dev/k8s.io/apimachinery@v0.26.1/pkg/util/validation/field would be better to describe the first situation.

Reply:
> /hold
> 
> There are merge commits in this PR, can we remove them please?

Sure, I would rebase them.

Reply:
We should at least introduce a struct, like suggested in the first comment? The error can be separate

Reply:
Fine for me. So the struct would only contain warnings, while errors should still be returned via the err field.

Reply:
What about something like:

```golang
type Warnings []string

[...]
ValidateDelete() (Warnings, error)
```

So that the return is clear and expressive, and simple to use?

Reply:
Sounds good to me!

Reply:
> What about something like:
> 
> ```go
> type Warnings []string
> 
> [...]
> ValidateDelete() (Warnings, error)
> ```
> 
> So that the return is clear and expressive, and simple to use?

I face some troubles when I refactoring `[]string` -> `type Warnings []string`, it's the import cycle between packages 

- `sigs.k8s.io/controller-runtime/pkg/webhook/admission` and 
- `sigs.k8s.io/controller-runtime/pkg/webhook/admission/admissiontest`

In detail, the second package introduces a "mock implementation" for the `Validator`, and the "mocked Validator" is used in the first package test codes.

Before, there was only one import path: `sigs.k8s.io/controller-runtime/pkg/webhook/admission`  -> `sigs.k8s.io/controller-runtime/pkg/webhook/admission/admissiontest`.

But after introducing the new struct `type Warnings []string`, it's necessary to use the new struct as the type of parameter to implement the `Validator` interface. Then the new `import` statement is needed in the second package, then the import cycle happens.

I'd prefer to make these 2 packages into one package, if A depends on B and also B depends on A, they are actually one thing. More accurately, move `pkg/webhook/admission/admissiontest/util.go` ->  `pkg/webhook/admission/mock_validator_test.go`(or as other filename)

What do you think about it? @vincepri @sbueringer :heart:

Reply:
I think it sounds reasonable to just move the FakeValidator into the admission package. I would "unexport" it and move it directly in validator_test.go

It looks like FakeValidator is not really used outside of controller-runtime: https://cs.k8s.io/?q=%22sigs.k8s.io%2Fcontroller-runtime%2Fpkg%2Fwebhook%2Fadmission%2Fadmissiontest%22&i=nope&files=&excludeFiles=&repos=

So moving it and making it internal should be fine. In case we think we should preserve the FakeValidator for external consumers I would suggest leaving the admissiontest package as is but still use an unexported local version in our own unit tests

Reply:
Get that! Will do!